src/libraries/System.Reflection.DispatchProxy/src/System/Reflection/DispatchProxyGenerator.cs

```diff
@@ -117,19 +117,9 @@ private sealed class ProxyAssembly
             [RequiresDynamicCode("Defining a dynamic assembly requires generating code at runtime")]
             public ProxyAssembly(AssemblyLoadContext alc)
             {
-                string name;
-                if (alc == AssemblyLoadContext.Default)
-                {
-                    name = "ProxyBuilder";
-                }
-                else
-                {
-                    string? alcName = alc.Name;
-                    name = string.IsNullOrEmpty(alcName) ? $"DispatchProxyTypes.{alc.GetHashCode()}" : $"DispatchProxyTypes.{alcName}";
```

Comment:
Should we keep the `GetHashCode()` path here?  That would preserve unique names and help with debuggability and could avoid folks coupling to this AssemblyName (which presumably we don't want).  Not meaning to reset this change but I couldn't see if this was considered when looking at https://github.com/dotnet/runtime/pull/82754.  I'm fine either way -- just want to hear folks thoughts.  cc @jkotas @vitek-karas

Reply:
The need for disambiguation of assemblies with same name loaded in different load contexts is not unique to DispatchProxy. It would be best to solve this by teaching debuggers and other diagnostic tools to display the assembly context name as needed. This was suggested e.g. in https://github.com/dotnet/runtime/issues/80387#issuecomment-1428370220.

I do not have a strong opinion about one-off improvements like including hashcode in the generated assembly name here.

Reply:
The problem is that the load context name is arbitrary, but using it in the assembly name means that certain characters can trigger the AssemblyNameParser to expect the remaining string to follow a particular pattern. I think removing whitespace plus the following characters would make the name 'safe'

, = : / \ 

So, perhaps just replace all those characters (plus whitespace) with underscores when constructing the assembly name?

Reply:
Assembly name can contain any characters (except `\0`). There is a difference between full name and simple name. The full name can have version, etc. and special characters have to be properly escaped. The simple name can contain any characters. The problem was caused by full name not being properly escaped.

One can convert between simple name and (properly escaped) full name using AssemblyType type.

```
using System.Reflection;

var an = new AssemblyName();
an.Name = "NameWithCharThatNeedEscaping=";
Console.WriteLine(an.ToString());
```

prints `NameWithCharThatNeedEscaping\=` that is a valid assembly name.

Reply:
So, just change the one line then to escape the name of the load context when building the assembly name?

`name = string.IsNullOrEmpty(alcName) ? $"DispatchProxyTypes.{alc.GetHashCode()}" : $"DispatchProxyTypes.{new AssemblyName() {Name = alcName}}";`

Reply:
Yes, this would get the name properly escaped.

Unfortunately, there is a second problem: `DefineDynamicAssembly` argument validation rejects valid assembly names: https://github.com/dotnet/runtime/blob/f8e6ebfd87da7d8947a8cf550cff9ec51f32cd1e/src/coreclr/vm/assembly.cpp#L399-L405 . We would also need to fix that to make these atypical assembly names work.

Reply:
Discussed about this with @jkotas offline: `the servicing fix looks good as is because it is a safe low-risk change. If we want to make this better, it should be in .NET 8 only`. Created an issue for the 8.0 fix https://github.com/dotnet/runtime/issues/82969